examples/typescript-firebase-todo/src/App.tsx

```diff
@@ -0,0 +1,359 @@
+    ref.on('child_added', on('add'))
+    ref.on('child_removed', on('retract'))
+    ref.on('child_changed', on('add'))
+    return () => {
+      client.removeTransactListener()
+      ref.off('child_added', on('add'))
+      ref.off('child_removed', on('retract'))
+      ref.off('child_changed', on('add'))
+    }
+  }, [userId])
+  return null
+}
+
+const SignOut = () => {
+  const [client] = useClient()
+  return (
+    <button 
+      style={{float: 'right'}} 
+      onClick={() => {
+        
+        client.dbFromString(window.emptyDB)
+        firebase.auth().signOut()
+      }}
+    >Sign Out</button>
+  )
+}
+
+const NewTodo = () => {
+  const [transact] = useTransact()
+  return (
+    <form onSubmit={(e:any) => {
+      e.preventDefault()
+      transact([{
+        todo: {
+          name: e.target.elements['todo-name'].value,
+          createdAt: Date.now()
+        }
+      }])
+      e.target.reset()
+    }}>
+      <input 
+        autoFocus
+        style={{fontSize: 20}}
+        type="text" 
+        name="todo-name" 
+        placeholder="What needs to be done?" 
+        autoComplete="off"
+        required
+      />
+      &nbsp;
+      <button type="submit">Create Todo</button>
+    </form>
+  )
+}
+
+const TodoList = () => {
+  const [filters] = useEntity({ identity: 'todoFilters' })
+  const [todos] = useQuery({
+    $find: 'todo',
+    $where: { todo: { name: '$any' } }
+  })
+  return (
+    <div>
+      {todos.filter(todo => {
+        if (!filters.get('showCompleted') && todo.get('isCompleted')) return false
+        if (filters.get('project') && todo.get('project', 'id') !== filters.get('project')) return false
+        if (filters.get('owner') && todo.get('owner', 'id') !== filters.get('owner')) return false
+        return true
+      }).sort((a, b) => a.get('createdAt') > b.get('createdAt') ? -1 : 1)
+      .map(todo => <Todo key={todo.get('id')} id={todo.get('id')}/>)}
+    </div>
+  )
+}
+
+// PERFORMANCE: By accepting an `id` prop instead of a whole `todo` entity
+// this component stays disconnected from the useQuery in the parent TodoList. 
+// useEntity creates a separate scope for every Todo so changes to TodoList
+// or sibling Todos don't trigger unnecessary re-renders.
+const Todo = React.memo(({ id } : {id : any}) => {
+  const [todo] = useEntity(id)
+  return (
+    <div>
+      <div style={{ display: 'flex', flexDirection: 'row', alignItems: 'flex-end', paddingTop: 20}}>
+        <TodoCheck todo={todo} />
+        <TodoName todo={todo} />
+      </div>
+      <div>
+        <TodoProject todo={todo} />
+        &nbsp;·&nbsp;
+        <TodoOwner todo={todo} />
+        &nbsp;·&nbsp;
+        <TodoDelete todo={todo} />
+      </div>
+      <small style={{ color: 'grey' }}>
+        {new Date(todo.get('createdAt')).toLocaleString()}
+      </small>
+    </div>
+  )
+})
+
+const TodoCheck = ({ todo } : {todo: any}) => {
```

Comment:
Is it possible to type todo as an Entity?

Reply:
yes done